File: api/src/main/java/com/ruby/rubia_server/core/controller/MessageEnhancementAuditController.java
package com.ruby.rubia_server.core.controller;

import com.ruby.rubia_server.core.entity.MessageEnhancementAudit;
import com.ruby.rubia_server.core.service.MessageEnhancementAuditService;
import com.ruby.rubia_server.core.util.CompanyContextUtil;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/api/message-enhancement-audit")
@RequiredArgsConstructor
@Slf4j
public class MessageEnhancementAuditController {

    private final MessageEnhancementAuditService auditService;
    private final CompanyContextUtil companyContextUtil;

    @GetMapping("/company/{companyId}")
    public ResponseEntity<Page<MessageEnhancementAudit>> getAuditsByCompany(
            @PathVariable UUID companyId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size,
            @RequestParam(defaultValue = "createdAt") String sortBy,
            @RequestParam(defaultValue = "desc") String sortDir) {

        log.debug("Fetching message enhancement audits for company: {}", companyId);

        // Validate company context
        companyContextUtil.ensureCompanyAccess(companyId);

        Sort sort = sortDir.equalsIgnoreCase("desc") ? 
            Sort.by(sortBy).descending() : Sort.by(sortBy).ascending();
        Pageable pageable = PageRequest.of(page, size, sort);

        Page<MessageEnhancementAudit> audits = auditService.getAuditsByCompany(companyId, pageable);
        return ResponseEntity.ok(audits);
    }

    @GetMapping("/company/{companyId}/stats")
    public ResponseEntity<MessageEnhancementAuditService.EnhancementStats> getCompanyStats(
            @PathVariable UUID companyId) {

        log.debug("Fetching enhancement stats for company: {}", companyId);

        // Validate company context
        companyContextUtil.ensureCompanyAccess(companyId);

        MessageEnhancementAuditService.EnhancementStats stats = auditService.getCompanyStats(companyId);
        return ResponseEntity.ok(stats);
    }

    @GetMapping("/company/{companyId}/by-date-range")
    public ResponseEntity<Page<MessageEnhancementAudit>> getAuditsByDateRange(
            @PathVariable UUID companyId,
            @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime startDate,
            @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime endDate,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {

        log.debug("Fetching audits for company: {} between {} and {}", companyId, startDate, endDate);

        // Validate company context
        companyContextUtil.ensureCompanyAccess(companyId);

        Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        Page<MessageEnhancementAudit> audits = auditService.getAuditsByDateRange(companyId, startDate, endDate, pageable);
        return ResponseEntity.ok(audits);
    }

    @GetMapping("/company/{companyId}/by-temperament/{temperament}")
    public ResponseEntity<Page<MessageEnhancementAudit>> getAuditsByTemperament(
            @PathVariable UUID companyId,
            @PathVariable String temperament,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {

        log.debug("Fetching audits for company: {} with temperament: {}", companyId, temperament);

        // Validate company context
        companyContextUtil.ensureCompanyAccess(companyId);

        Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        Page<MessageEnhancementAudit> audits = auditService.getAuditsByTemperament(companyId, temperament, pageable);
        return ResponseEntity.ok(audits);
    }

    @GetMapping("/company/{companyId}/by-ai-model/{model}")
    public ResponseEntity<Page<MessageEnhancementAudit>> getAuditsByAiModel(
            @PathVariable UUID companyId,
            @PathVariable String model,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {

        log.debug("Fetching audits for company: {} with AI model: {}", companyId, model);

        // Validate company context
        companyContextUtil.ensureCompanyAccess(companyId);

        Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        Page<MessageEnhancementAudit> audits = auditService.getAuditsByAiModel(companyId, model, pageable);
        return ResponseEntity.ok(audits);
    }

    @GetMapping("/user/{userId}")
    public ResponseEntity<Page<MessageEnhancementAudit>> getAuditsByUser(
            @PathVariable UUID userId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {

        log.debug("Fetching audits for user: {}", userId);

        Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        Page<MessageEnhancementAudit> audits = auditService.getAuditsByUser(userId, pageable);
        return ResponseEntity.ok(audits);
    }

    @GetMapping("/ai-agent/{aiAgentId}")
    public ResponseEntity<Page<MessageEnhancementAudit>> getAuditsByAiAgent(
            @PathVariable UUID aiAgentId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {

        log.debug("Fetching audits for AI agent: {}", aiAgentId);

        Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        Page<MessageEnhancementAudit> audits = auditService.getAuditsByAiAgent(aiAgentId, pageable);
        return ResponseEntity.ok(audits);
    }

    @GetMapping("/conversation/{conversationId}")
    public ResponseEntity<List<MessageEnhancementAudit>> getAuditsByConversation(
            @PathVariable UUID conversationId) {

        log.debug("Fetching audits for conversation: {}", conversationId);

        List<MessageEnhancementAudit> audits = auditService.getAuditsByConversation(conversationId);
        return ResponseEntity.ok(audits);
    }
}

File: api-dotnet/Rubia.Server/Controllers/MessageEnhancementAuditController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Rubia.Server.Entities;
using Rubia.Server.Services;
using System.ComponentModel.DataAnnotations;

namespace Rubia.Server.Controllers;

[ApiController]
[Route("api/message-enhancement-audit")]
[Authorize]
public class MessageEnhancementAuditController : ControllerBase
{
    private readonly MessageEnhancementAuditService _auditService;
    private readonly CompanyContextService _companyContextService;
    private readonly ILogger<MessageEnhancementAuditController> _logger;

    public MessageEnhancementAuditController(
        MessageEnhancementAuditService auditService,
        CompanyContextService companyContextService,
        ILogger<MessageEnhancementAuditController> logger)
    {
        _auditService = auditService;
        _companyContextService = companyContextService;
        _logger = logger;
    }

    [HttpGet("company/{companyId}")]
    public async Task<ActionResult<PagedResult<MessageEnhancementAudit>>> GetAuditsByCompany(
        Guid companyId,
        [FromQuery] int page = 0,
        [FromQuery] int size = 20,
        [FromQuery] string sortBy = "createdAt",
        [FromQuery] string sortDir = "desc")
    {
        _logger.LogDebug("Fetching message enhancement audits for company: {CompanyId}", companyId);

        await _companyContextService.EnsureCompanyAccessAsync(companyId);

        var (items, totalCount) = await _auditService.GetAuditsByCompanyAsync(companyId, page, size);

        return Ok(new PagedResult<MessageEnhancementAudit>
        {
            Items = items,
            TotalCount = totalCount,
            Page = page,
            Size = size
        });
    }

    [HttpGet("company/{companyId}/stats")]
    public async Task<ActionResult<EnhancementStats>> GetCompanyStats(Guid companyId)
    {
        _logger.LogDebug("Fetching enhancement stats for company: {CompanyId}", companyId);

        await _companyContextService.EnsureCompanyAccessAsync(companyId);

        var stats = await _auditService.GetCompanyStatsAsync(companyId);
        return Ok(stats);
    }

    [HttpGet("company/{companyId}/by-date-range")]
    public async Task<ActionResult<PagedResult<MessageEnhancementAudit>>> GetAuditsByDateRange(
        Guid companyId,
        [FromQuery, Required] DateTime startDate,
        [FromQuery, Required] DateTime endDate,
        [FromQuery] int page = 0,
        [FromQuery] int size = 20)
    {
        _logger.LogDebug("Fetching audits for company: {CompanyId} between {StartDate} and {EndDate}",
            companyId, startDate, endDate);

        await _companyContextService.EnsureCompanyAccessAsync(companyId);

        var (items, totalCount) = await _auditService.GetAuditsByDateRangeAsync(
            companyId, startDate, endDate, page, size);

        return Ok(new PagedResult<MessageEnhancementAudit>
        {
            Items = items,
            TotalCount = totalCount,
            Page = page,
            Size = size
        });
    }

    [HttpGet("company/{companyId}/by-temperament/{temperament}")]
    public async Task<ActionResult<PagedResult<MessageEnhancementAudit>>> GetAuditsByTemperament(
        Guid companyId,
        string temperament,
        [FromQuery] int page = 0,
        [FromQuery] int size = 20)
    {
        _logger.LogDebug("Fetching audits for company: {CompanyId} with temperament: {Temperament}",
            companyId, temperament);

        await _companyContextService.EnsureCompanyAccessAsync(companyId);

        var (items, totalCount) = await _auditService.GetAuditsByTemperamentAsync(
            companyId, temperament, page, size);

        return Ok(new PagedResult<MessageEnhancementAudit>
        {
            Items = items,
            TotalCount = totalCount,
            Page = page,
            Size = size
        });
    }

    [HttpGet("company/{companyId}/by-ai-model/{model}")]
    public async Task<ActionResult<PagedResult<MessageEnhancementAudit>>> GetAuditsByAiModel(
        Guid companyId,
        string model,
        [FromQuery] int page = 0,
        [FromQuery] int size = 20)
    {
        _logger.LogDebug("Fetching audits for company: {CompanyId} with AI model: {Model}",
            companyId, model);

        await _companyContextService.EnsureCompanyAccessAsync(companyId);

        var (items, totalCount) = await _auditService.GetAuditsByAiModelAsync(
            companyId, model, page, size);

        return Ok(new PagedResult<MessageEnhancementAudit>
        {
            Items = items,
            TotalCount = totalCount,
            Page = page,
            Size = size
        });
    }

    [HttpGet("user/{userId}")]
    public async Task<ActionResult<PagedResult<MessageEnhancementAudit>>> GetAuditsByUser(
        Guid userId,
        [FromQuery] int page = 0,
        [FromQuery] int size = 20)
    {
        _logger.LogDebug("Fetching audits for user: {UserId}", userId);

        var (items, totalCount) = await _auditService.GetAuditsByUserAsync(userId, page, size);

        return Ok(new PagedResult<MessageEnhancementAudit>
        {
            Items = items,
            TotalCount = totalCount,
            Page = page,
            Size = size
        });
    }

    [HttpGet("ai-agent/{aiAgentId}")]
    public async Task<ActionResult<PagedResult<MessageEnhancementAudit>>> GetAuditsByAiAgent(
        Guid aiAgentId,
        [FromQuery] int page = 0,
        [FromQuery] int size = 20)
    {
        _logger.LogDebug("Fetching audits for AI agent: {AiAgentId}", aiAgentId);

        var (items, totalCount) = await _auditService.GetAuditsByAiAgentAsync(aiAgentId, page, size);

        return Ok(new PagedResult<MessageEnhancementAudit>
        {
            Items = items,
            TotalCount = totalCount,
            Page = page,
            Size = size
        });
    }

    [HttpGet("conversation/{conversationId}")]
    public async Task<ActionResult<List<MessageEnhancementAudit>>> GetAuditsByConversation(
        Guid conversationId)
    {
        _logger.LogDebug("Fetching audits for conversation: {ConversationId}", conversationId);

        var audits = await _auditService.GetAuditsByConversationAsync(conversationId);
        return Ok(audits);
    }
}

public class PagedResult<T>
{
    public List<T> Items { get; set; } = [];
    public int TotalCount { get; set; }
    public int Page { get; set; }
    public int Size { get; set; }
    public int TotalPages => (int)Math.Ceiling((double)TotalCount / Size);
}